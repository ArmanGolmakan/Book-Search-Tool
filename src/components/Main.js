import React, { useState, useEffect } from 'react';
import './Main.style.css';
import Card from './Card.js';

const Main = () => {

    const [data, setData] = useState({ items: [] });
    const [bookCards, setBookCards] = useState([]);
    const [searchText, setSearchText] = useState('');
    const [error, setError] = useState('');
    const [loading, setLoading] = useState(false);
    const [order, setOrder] = useState('');

    useEffect(() => {
        let content = [];
        if (data.docs) {
            let sortedBooks = data.docs;
            if (order == 'alphabetical') {
                sortedBooks.sort((a, b) => (a.title > b.title) ? 1 : -1);
            }
            else if (order == 'publishdate') {
                console.log("XXXXXXXXXXXX");
                sortedBooks.sort((a, b) => (a.first_publish_year > b.first_publish_year) ? 1 : -1);
            }
            sortedBooks.map((item, index) => {
                content.push(
                    <li key='index'>
                        <Card bookData={item}></Card>
                    </li>
                );
            });
            setBookCards(content);
        }
    }, [data, order]);

    const handleSubmit = async (event) => {
        setLoading(true);
        event.preventDefault(); //prevents refreshing webpage
        const myData = await fetch("https://openlibrary.org/search.json?q=" + searchText);
        const myDataJson = await myData.json();
        setData(myDataJson);
        setLoading(false);
    };

    const BooksSection = () => {
        return (
            <div>
                <ContentHead />
                {bookCards}
            </div>
        )
    }

    const sortByTitle = () => {
        setOrder('alphabetical');
    }

    const sortByPublishDate = () => {
        setOrder('publishdate');
    }

    const ContentHead = () => {
        return (
            <div className='contentHead'>
                <span>{data.num_found} results </span>
                <span>Sort by: <button onClick={sortByTitle}>alphabetical</button> |
                    <button onClick={sortByPublishDate}>publish date</button></span>
            </div>
        )
    }

    return (
        <div className='parentContainer'>
            <h1>Book Search Tool</h1>
            <div className='search-field-container'>
                <form onSubmit={handleSubmit}>
                    <input type='search' placeholder='Search...' onChange={e => setSearchText(e.target.value)} value={searchText} />
                    <button type='submit'>Search</button>
                </form>
            </div>

            {
                loading ? (
                    <div>
                        <h1>
                            Loading...
                        </h1>
                    </div>
                ) : (<></>)
            }
            {
                data.docs ? (
                    <div className='dataContainer'>
                        <BooksSection />
                    </div>
                ) : (<></>)
            }
        </div>
    );
}

export default Main;