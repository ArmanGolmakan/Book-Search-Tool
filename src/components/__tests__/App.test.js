import React from 'react';
import { shallow } from 'enzyme';
import '../../setupTest';
import App from "../../App";
import Main from "../Main";
import toJson from 'enzyme-to-json';

describe('rendering components', () => {
    //shallow only renders a single components. does NOT render its child components
    it('renders App without crashing', () => {
        shallow(<App />);
    });

    it('renders Main without crashing', () => {
        shallow(<Main />);
    });

    it('renders page header: Book Search Tool', () => {
        const wrapper = shallow(<Main />);
        const h2 = <h1>Book Search Tool</h1>;
        expect(wrapper.contains(h2)).toEqual(true);
    });

    it('renders search button', () => {
        const wrapper = shallow(<Main />);
        const button = <button type='submit'>Search</button>;
        expect(wrapper.contains(button)).toEqual(true);
    })
})

describe('snapshots', () => {
    it ('App snapshots', () => {
        const appTree = shallow(<App/>);
        expect(toJson(appTree)).toMatchSnapshot();
    })
    it ('Main snapshots', () => {
        const mainTree = shallow(<Main/>);
        expect(toJson(mainTree)).toMatchSnapshot();
    })
})



