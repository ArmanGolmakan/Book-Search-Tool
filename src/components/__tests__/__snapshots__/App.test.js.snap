// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots App snapshots 1`] = `
<div
  className="App"
>
  <Main />
</div>
`;

exports[`snapshots Main snapshots 1`] = `
<div
  className="parentContainer"
>
  <h1>
    Book Search Tool
  </h1>
  <div
    className="search-field-container"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Search..."
        type="search"
        value=""
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</div>
`;
